File: aoc23/days/09/scala/09.scala
import scala.io.Source

object run {
    def main(args: Array[String]) = {
        val source = Source.fromFile("../input")
        val (l, r) = (for line <- source.getLines() yield solve(parse(line))).toArray.unzip
        println(f"Part one: ${r.sum}, Part two: ${l.sum} ")
        source.close()
    }

    def parse(line: String): Array[Int] = line.split(" ").map(i => i.toInt)

    def solve(ns : Array[Int]): (Int, Int) = if (ns.forall(_ == 0)) (0, 0) else {
        val (l, r) = solve(ns.zip(ns.tail).map((i, j) => j - i))
        return (ns.head - l, ns.last + r)
    }
}

File: aoc23/days/09/09.py
with open("input") as file:
    raw = file.read()

lines = raw.strip().split("\n")
p1 = p2 = 0

for line in lines:
    diffs = [[int(x) for x in line.split()]]
    while not all(i == 0 for i in diffs[-1]):
        cur = diffs[-1]
        diffs.append([j - i for i, j in zip(cur[:-1], cur[1:])])

    p1 += sum([x[-1] for x in diffs])
    wrk = 0
    for diff in diffs[::-1]:
        wrk = diff[0] - wrk
    p2 += wrk

print(p1, p2)
